openshift/openshift.go

```diff
@@ -0,0 +1,176 @@
+		},
+	}
+
+	configMaps := client.CoreV1().ConfigMaps(ns)
+
+	nameSelector := fields.OneTermEqualSelector("metadata.name", cfgMapName).String()
+	listOpts := metav1.ListOptions{
+		Watch:         true,
+		FieldSelector: nameSelector,
+	}
+
+	watch, err := configMaps.Watch(ctx, listOpts)
+	if err != nil {
+		return nil, err
+	}
+	defer watch.Stop()
+
+	crtChan := make(chan string)
+	go func() {
+		for event := range watch.ResultChan() {
+			cm, ok := event.Object.(*v1.ConfigMap)
+			if !ok {
+				continue
+			}
+			if crt, ok := cm.Data["service-ca.crt"]; ok {
+				crtChan <- crt
+				close(crtChan)
+				break
+			}
+		}
+	}()
+
+	_, err = configMaps.Create(ctx, cfgMap, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = configMaps.Delete(ctx, cfgMapName, metav1.DeleteOptions{})
+	}()
+
+	select {
+	case crt := <-crtChan:
+		blk, _ := pem.Decode([]byte(crt))
+		return x509.ParseCertificate(blk.Bytes)
+	case <-time.After(time.Second * 5):
+		return nil, errors.New("failed to get OpenShift's service CA in time")
+	}
+}
+
+// WithOpenShiftServiceCA enables trust to OpenShift's service CA for internal image registry
+func WithOpenShiftServiceCA() fnhttp.Option {
+	if !IsOpenShift() {
+		return nil
+	}
+
+	var selectCA func(ctx context.Context, serverName string) (*x509.Certificate, error)
+	ca, err := GetServiceCA(context.TODO())
+	if err == nil {
+		selectCA = func(ctx context.Context, serverName string) (*x509.Certificate, error) {
+			if strings.HasPrefix(serverName, registryHost) {
+				return ca, nil
+			}
+			return nil, nil
+		}
+	}
+	return fnhttp.WithSelectCA(selectCA)
+}
+
+func GetDefaultRegistry() string {
+	if !IsOpenShift() {
+		return ""
+	}
+
+	ns, _ := k8s.GetNamespace("")
+	if ns == "" {
+		ns = "default"
+	}
+
+	return registryHostPort + "/" + ns
+}
+
+func GetDockerCredentialLoaders() []creds.CredentialsCallback {
+	if !IsOpenShift() {
+		return nil
+	}
+
+	conf := k8s.GetClientConfig()
+
+	rawConf, err := conf.RawConfig()
+	if err != nil {
+		return nil
+	}
+
+	cc := rawConf.Contexts[rawConf.CurrentContext]
+	authInfo := rawConf.AuthInfos[cc.AuthInfo]
+
+	user := "user"
+	parts := strings.SplitN(cc.AuthInfo, "/", 2)
+	if len(parts) >= 1 {
+		user = parts[0]
+	}
+	credentials := docker.Credentials{
+		Username: user,
+		Password: authInfo.Token,
+	}
+
+	return []creds.CredentialsCallback{
+		func(registry string) (docker.Credentials, error) {
+			if registry == registryHostPort {
+				return credentials, nil
+			}
+			return docker.Credentials{}, nil
+		},
+	}
+
+}
+
+var isOpenShift bool
+var checkOpenShiftOnce sync.Once
+
+func IsOpenShift() bool {
+	checkOpenShiftOnce.Do(func() {
+		client, err := k8s.NewKubernetesClientset()
+		if err != nil {
+			isOpenShift = false
+			return
+		}
+		_, err = client.CoreV1().Services("openshift-image-registry").Get(context.TODO(), "image-registry", metav1.GetOptions{})
```

Comment:
Hmm -- this will return `false` if the user doesn't have `services` `get` permission on the `openshift-image-registry` namespace.

I'm not sure if there's another / better way to detect this, either from server headers or by probing the registry and looking at its header / challenge. I think `go-containerregistry` uses a `GET /v2/ HTTP /1.1` probe of the registry to figure out what and whether auth is required.

Reply:
https://github.com/google/go-containerregistry/tree/main/pkg/authn#the-registry suggests looking at the `Www-Authenticate` header for clues as to the authentication mechanism.

Reply:
I would avoid doing HTTP request to internal service here -- it requires some setup I don't want to do here.

Reply:
>Hmm -- this will return false if the user doesn't have services get permission on the openshift-image-registry namespace.

@evankanderson  This does work with non-privileged user. Non-privilieged user will get different errors for non-existent service and for service that is not just permitted.